src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -337,87 +354,279 @@ out ImmutableArray<(BoundExpression, BoundDagTemp)> bindings)
             }
         }
 
-        private static BoundDecisionDag MakeDecisionDag(ImmutableArray<PartialCaseDecision> cases)
+        private BoundDecisionDag MakeDecisionDag(SyntaxNode syntax, ImmutableArray<PartialCaseDecision> cases, LabelSymbol defaultLabel)
         {
-            throw new NotImplementedException();
-        }
-    }
+            if (cases.IsDefaultOrEmpty)
+            {
+                return new BoundDecision(syntax, defaultLabel);
+            }
 
-    internal class PartialCaseDecision
-    {
-        public readonly int Index;
-        public readonly ImmutableArray<BoundDagDecision> Decisions;
-        public readonly ImmutableArray<(BoundExpression, BoundDagTemp)> Bindings;
-        public readonly BoundExpression WhereClause;
-        public readonly LabelSymbol CaseLabel;
-        public PartialCaseDecision(
-            int Index,
-            ImmutableArray<BoundDagDecision> Decisions,
-            ImmutableArray<(BoundExpression, BoundDagTemp)> Bindings,
-            BoundExpression WhereClause,
-            LabelSymbol CaseLabel)
-        {
-            this.Index = Index;
-            this.Decisions = Decisions;
-            this.Bindings = Bindings;
-            this.WhereClause = WhereClause;
-            this.CaseLabel = CaseLabel;
+            var first = cases[0];
+            if (first.Decisions.IsDefaultOrEmpty)
+            {
+                // The first pattern has fully matched
+                if (first.WhereClause == null || first.WhereClause.ConstantValue == ConstantValue.True)
+                {
+                    // no (or constant true) where clause. The decision is finalized.
+                    if (first.Bindings.IsDefaultOrEmpty)
+                    {
+                        return new BoundDecision(first.Syntax, first.CaseLabel);
+                    }
+                    else
+                    {
+                        return new BoundWhereClause(first.Syntax, first.Bindings, null, new BoundDecision(first.Syntax, first.CaseLabel), null);
+                    }
+                }
+                else
+                {
+                    // in case the where clause fails, we prepare for the remaining cases.
+                    var remainingCases = cases.WhereAsArray(d => d != first);
```

Comment:
>WhereAsArray(d => d != first) [](start = 47, length = 29)

Is this equivalent to Skip(1), or RemoveAt(0)? Consider using it instead.

Reply:
Unless array can contain duplicates.

---
In reply to: [159783678](https://github.com/dotnet/roslyn/pull/23228#discussion_r159783678) [](ancestors = 159783678)

Reply:
ImmutableArray does not have Skip or RemoveAt methods.

Reply:
> ImmutableArray does not have Skip or RemoveAt methods.

```
        //
        // Summary:
        //     Returns an array with the element at the specified position removed.
        //
        // Parameters:
        //   index:
        //     The 0-based index into the array for the element to omit from the returned array.
        //
        // Returns:
        //     The new array.
        public ImmutableArray<T> RemoveAt(int index);
```

Reply:
Oh, don't know why completion didn't see it. Addressed in https://github.com/dotnet/roslyn/pull/24059/commits/a6b45385606adad7241c8edf6997c3e19446d6e0 of https://github.com/dotnet/roslyn/pull/24059.